misk/src/main/kotlin/misk/clustering/DefaultCluster.kt

```diff
@@ -0,0 +1,156 @@
+import com.google.inject.Key
+import misk.DependentService
+import misk.inject.keyOf
+import misk.logging.getLogger
+import org.eclipse.jetty.util.BlockingArrayQueue
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+import javax.inject.Singleton
+
+/**
+ * A [DefaultCluster] is the default implementation of the [Cluster], which relies on an outside
+ * source such as a cluster watcher to inform it as members become ready or not ready. The
+ * [DefaultCluster] handles computing the actual cluster changes, and managing application watches
+ * that need to be triggered as the cluster changes.
+ */
+@Singleton
+internal class DefaultCluster(
+  self: Cluster.Member,
+  private val newResourceMapperFn: (members: Set<Cluster.Member>) -> ClusterResourceMapper =
+      { ClusterHashRing(it) }
+) : AbstractExecutionThreadService(), Cluster, DependentService {
+  private val snapshotRef = AtomicReference<Cluster.Snapshot>(Cluster.Snapshot(
+      self = self,
+      selfReady = false,
+      readyMembers = setOf(),
+      resourceMapper = newResourceMapper(setOf())
+  ))
+  private val running = AtomicBoolean(false)
+  private val actions = BlockingArrayQueue<(MutableSet<ClusterWatch>) -> Unit>()
+
+  override val snapshot: Cluster.Snapshot get() = snapshotRef.get()
+  override val consumedKeys: Set<Key<*>> = setOf()
+  override val producedKeys: Set<Key<*>> = setOf(keyOf<Cluster>())
+
+  /**
+   * Runs the internal event loop that handles requests to add watches or cluster changes. We use
+   * a single threaded event loop to ensure that watches get consistent diffs for the cluster
+   * without being racy or requiring locks that span application code (which might result in
+   * deadlocks). When a watch is registered, it is provided with the cluster membership as it
+   * is known at that time, and then receives diffs as they arrive.
+   */
+  override fun run() {
+    val watches = mutableSetOf<ClusterWatch>()
+
+    while (running.get()) {
+      try {
+        val action = actions.take()
+        action(watches)
+      } catch (th: Throwable) {
+        log.error(th) { "error calling event queue action" }
+      }
+    }
+  }
+
+  public override fun startUp() {
+    log.info { "starting cluster manager" }
+    running.set(true)
+  }
+
+  public override fun shutDown() {
+    log.info { "shutting down cluster manager" }
+    running.set(false)
+    actions.add { _ -> } // empty action to up the action queue if it is empty
+  }
+
+  /** Adds a new cluster watch */
+  override fun watch(watch: ClusterWatch) {
+    actions.add { watches ->
+      watches.add(watch)
+      watch(Cluster.Changes(snapshot))
+    }
+  }
+
+  /** Triggers a change to the cluster in response members becoming ready or not ready */
+  fun clusterChanged(
+    membersBecomingReady: Set<Cluster.Member> = setOf(),
+    membersBecomingNotReady: Set<Cluster.Member> = setOf()
+  ) {
+    actions.add { watches ->
```

Comment:
I really like how you’ve structured this . . . it’s all behavior and very little ceremony